@tinysun212 Sorry!  CI hit an error:

```
swift/validation-test/stdlib/Set.swift:108:21: error: missing argument label 'exclusiveUpperBound:' in call
  return Int(rand32(UInt32(max)))
                    ^
                    exclusiveUpperBound: 
```

--- a/stdlib/private/SwiftPrivate/PRNG.swift
+++ b/stdlib/private/SwiftPrivate/PRNG.swift
@@ -12,21 +12,18 @@
 
 import SwiftShims
 
-@inline(never) @_semantics("stdlib_binary_only") // Hide the libbsd dependency
 public func rand32() -> UInt32 {
-  return _swift_stdlib_arc4random()
+  return _swift_stdlib_cxx11_mt19937()
 }
 
-@inline(never) @_semantics("stdlib_binary_only") // Hide the libbsd dependency
 public func rand32(exclusiveUpperBound limit: UInt32) -> UInt32 {
-  return _swift_stdlib_arc4random_uniform(limit)
+  return _swift_stdlib_cxx11_mt19937_uniform(limit)
 }
 
-@inline(never) @_semantics("stdlib_binary_only") // Hide the libbsd dependency
 public func rand64() -> UInt64 {
   return
-    (UInt64(_swift_stdlib_arc4random()) << 32) |
-    UInt64(_swift_stdlib_arc4random())
+    (UInt64(_swift_stdlib_cxx11_mt19937()) << 32) |
+    UInt64(_swift_stdlib_cxx11_mt19937())
 }
 
 public func randInt() -> Int {

--- a/stdlib/public/SwiftShims/LibcShims.h
+++ b/stdlib/public/SwiftShims/LibcShims.h
@@ -69,10 +69,13 @@ int _swift_stdlib_close(int fd);
 __attribute__((const))
 SWIFT_RUNTIME_STDLIB_INTERFACE
 __swift_size_t _swift_stdlib_malloc_size(const void *ptr);
+
+// Random number <random>
 SWIFT_RUNTIME_STDLIB_INTERFACE
-__swift_uint32_t _swift_stdlib_arc4random(void);
+__swift_uint32_t _swift_stdlib_cxx11_mt19937(void);
 SWIFT_RUNTIME_STDLIB_INTERFACE
-__swift_uint32_t _swift_stdlib_arc4random_uniform(__swift_uint32_t upper_bound);
+__swift_uint32_t
+_swift_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound);
 
 #ifdef __cplusplus
 }} // extern "C", namespace swift

--- a/stdlib/public/stubs/LibcShims.cpp
+++ b/stdlib/public/stubs/LibcShims.cpp
@@ -10,6 +10,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include <random>
 #include <type_traits>
 #include <unistd.h>
 #include <stdlib.h>
@@ -69,11 +70,19 @@ size_t _swift_stdlib_malloc_size(const void *ptr) {
 #error No malloc_size analog known for this platform/libc.
 #endif
 
-__swift_uint32_t _swift_stdlib_arc4random(void) { return arc4random(); }
+static std::random_device RandomeDevice;
+static std::mt19937 MersenneRandom(RandomeDevice());
+
+__swift_uint32_t _swift_stdlib_cxx11_mt19937(void) {
+  return MersenneRandom();
+}
 
 __swift_uint32_t
-_swift_stdlib_arc4random_uniform(__swift_uint32_t upper_bound) {
-  return arc4random_uniform(upper_bound);
+_swift_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
+  if (upper_bound > 0)
+    upper_bound--;
+  std::uniform_int_distribution<__swift_uint32_t> RandomUniform(0, upper_bound);
+  return RandomUniform(MersenneRandom);
 }
 
 } // namespace swift

--- a/validation-test/stdlib/Set.swift
+++ b/validation-test/stdlib/Set.swift
@@ -105,7 +105,7 @@ func helperDeleteThree(k1: TestKeyTy, _ k2: TestKeyTy, _ k3: TestKeyTy) {
 }
 
 func uniformRandom(max: Int) -> Int {
-  return Int(arc4random_uniform(UInt32(max)))
+  return Int(rand32(exclusiveUpperBound: UInt32(max)))
 }
 
 func pickRandom<T>(a: [T]) -> T {

